{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import pysam\n",
    "import os\n",
    "from collections import defaultdict\n",
    "import sys\n",
    "from get_pairs import *\n",
    "\n",
    "directory = '/scratch/groups/dpwall/personal/chloehe/unmapped_reads/bam'\n",
    "pd.set_option('display.max_rows', 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fam 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# def pairs_generator(file, refname=None):\n",
    "    \n",
    "#     read_dict = defaultdict(lambda: [None, None])\n",
    "#     file.reset()\n",
    "#     for read in file.fetch(refname):\n",
    "#         qname = read.query_name\n",
    "#         if qname in read_dict:\n",
    "#             if read.is_read1:\n",
    "#                 yield qname, read, read_dict[qname][1]\n",
    "#             else:\n",
    "#                 yield qname, read_dict[qname][0], read\n",
    "#             del read_dict[qname]\n",
    "#         else:\n",
    "#             if read.is_read1:\n",
    "#                 read_dict[qname][0] = read\n",
    "#             else:\n",
    "#                 read_dict[qname][1] = read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_pairs_from_pairend_alignment(directory, refname=None):\n",
    "    \n",
    "    # check for alignment file and index file\n",
    "    try:\n",
    "        filename = glob.glob(os.path.join(directory, '*final.paired.aln_all.bam'))[0]\n",
    "    except:\n",
    "        print('Paired end alignment file not found!\\nExiting from get_pairs_from_pairend_alignment')\n",
    "        exit()\n",
    "    if not os.path.isfile(filename+'.bai'):\n",
    "        pysam.samtools.index(filename)\n",
    "    else:\n",
    "        if os.path.getctime(filename+'.bai') < os.path.getctime(filename):\n",
    "            pysam.samtools.index(filename)\n",
    "    \n",
    "    output_dict = dict()\n",
    "    read_dict = defaultdict(lambda: [None, None])\n",
    "    \n",
    "    alignmentfile = pysam.AlignmentFile(filename, 'rb')\n",
    "    alignmentfile.reset()\n",
    "    for read in alignmentfile.fetch(refname, until_eof=True):\n",
    "        qname = read.query_name\n",
    "        if qname in read_dict:\n",
    "            # sanity check: no qname should have been in output_dict already\n",
    "            if qname in output_dict:\n",
    "                print('Duplicate read warning: {}, skip and continue'.format(qname))\n",
    "                continue\n",
    "            if read.is_unmapped:\n",
    "                if read.is_read1:\n",
    "                    r2 = read_dict[qname][1]\n",
    "                    if r2 is 'unmapped':\n",
    "                        output_dict[qname] = ['unmapped', 'unmapped', 'N/A',\n",
    "                                              r2, r2, 'N/A',\n",
    "                                              'False']\n",
    "                    else:\n",
    "                        output_dict[qname] = ['unmapped', 'unmapped', 'N/A',\n",
    "                                              alignmentfile.get_reference_name(r2.reference_id),\n",
    "                                              r2.reference_start,\n",
    "                                              r2.mapping_quality,\n",
    "                                              'False']\n",
    "                    \n",
    "                else:\n",
    "                    r1 = read_dict[qname][0]\n",
    "                    if r1 is 'unmapped':\n",
    "                        output_dict[qname] = [r1, r1, 'N/A',\n",
    "                                              'unmapped', 'unmapped', 'N/A',\n",
    "                                              'False']\n",
    "                    else:\n",
    "                        output_dict[qname] = [alignmentfile.get_reference_name(r1.reference_id),\n",
    "                                              r1.reference_start,\n",
    "                                              r1.mapping_quality,\n",
    "                                              'unmapped', 'unmapped', 'N/A',\n",
    "                                              'False']\n",
    "            else:\n",
    "                if read.is_read1:\n",
    "                    r2 = read_dict[qname][1]\n",
    "                    if r2 is 'unmapped':\n",
    "                        output_dict[qname] = [alignmentfile.get_reference_name(read.reference_id),\n",
    "                                              read.reference_start,\n",
    "                                              read.mapping_quality,\n",
    "                                              r2, r2, 'N/A',\n",
    "                                              'False']\n",
    "                    else:\n",
    "                        output_dict[qname] = [alignmentfile.get_reference_name(read.reference_id),\n",
    "                                              read.reference_start,\n",
    "                                              read.mapping_quality,\n",
    "                                              alignmentfile.get_reference_name(r2.reference_id),\n",
    "                                              r2.reference_start,\n",
    "                                              r2.mapping_quality,\n",
    "                                              read.is_proper_pair]\n",
    "                    \n",
    "                else:\n",
    "                    r1 = read_dict[qname][0]\n",
    "                    if r1 is 'unmapped':\n",
    "                        output_dict[qname] = [r1, r1, 'N/A',\n",
    "                                              alignmentfile.get_reference_name(read.reference_id),\n",
    "                                              read.reference_start,\n",
    "                                              read.mapping_quality,\n",
    "                                              'False']\n",
    "                    else:\n",
    "                        output_dict[qname] = [alignmentfile.get_reference_name(r1.reference_id),\n",
    "                                              r1.reference_start,\n",
    "                                              r1.mapping_quality,\n",
    "                                              alignmentfile.get_reference_name(read.reference_id),\n",
    "                                              read.reference_start,\n",
    "                                              read.mapping_quality,\n",
    "                                              r1.is_proper_pair]\n",
    "            # delete entry from dictionary as soon as its mate is found\n",
    "            del read_dict[qname]\n",
    "        else:\n",
    "            if read.is_unmapped:\n",
    "                if read.is_read1:\n",
    "                    read_dict[qname][0] = 'unmapped'\n",
    "                else:\n",
    "                    read_dict[qname][1] = 'unmapped'\n",
    "            else:\n",
    "                if read.is_read1:\n",
    "                    read_dict[qname][0] = read\n",
    "                else:\n",
    "                    read_dict[qname][1] = read\n",
    "\n",
    "    # sanity check: read_dict should be empty if all single-end reads found their mates\n",
    "    if bool(read_dict):\n",
    "        print('Warning: pairing error!')\n",
    "        for qname in read_dict:\n",
    "            print('Mate is not found for '+qname)\n",
    "        print('Skip and continue')\n",
    "        \n",
    "    alignmentfile.close()\n",
    "    print('Done processing paired-end alignment file!')\n",
    "    \n",
    "    return output_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done processing paired-end alignment file!\n"
     ]
    }
   ],
   "source": [
    "paired_alignment_dict = get_pairs_from_pairend_alignment(os.path.join(directory, 'fam1/MH0143018'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>R1_ref</th>\n",
       "      <th>R1_start</th>\n",
       "      <th>R1_MAPQ</th>\n",
       "      <th>R2_ref</th>\n",
       "      <th>R2_start</th>\n",
       "      <th>R2_MAPQ</th>\n",
       "      <th>is_proper_pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:4:2117:4400:26062</th>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>316</td>\n",
       "      <td>60</td>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>38</td>\n",
       "      <td>60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:6:1105:16083:63420</th>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>325</td>\n",
       "      <td>50</td>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>86</td>\n",
       "      <td>40</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:6:1116:6257:25025</th>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>432</td>\n",
       "      <td>60</td>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>195</td>\n",
       "      <td>60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:2:1201:2299:47352</th>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>43</td>\n",
       "      <td>60</td>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>438</td>\n",
       "      <td>60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:4:1213:2238:67780</th>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>454</td>\n",
       "      <td>60</td>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>254</td>\n",
       "      <td>60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:8:2224:32505:25939</th>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:8:2224:32526:4596</th>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:8:2224:32546:28787</th>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:8:2224:32627:67006</th>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:8:2224:32820:66883</th>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4025201 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             R1_ref  R1_start R1_MAPQ  \\\n",
       "E00218:136:HC35MCCXX:4:2117:4400:26062   MH533023.1       316      60   \n",
       "E00218:136:HC35MCCXX:6:1105:16083:63420  MH533023.1       325      50   \n",
       "E00218:136:HC35MCCXX:6:1116:6257:25025   MH533023.1       432      60   \n",
       "E00218:136:HC35MCCXX:2:1201:2299:47352   MH533023.1        43      60   \n",
       "E00218:136:HC35MCCXX:4:1213:2238:67780   MH533023.1       454      60   \n",
       "...                                             ...       ...     ...   \n",
       "E00218:136:HC35MCCXX:8:2224:32505:25939    unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:8:2224:32526:4596     unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:8:2224:32546:28787    unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:8:2224:32627:67006    unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:8:2224:32820:66883    unmapped  unmapped     N/A   \n",
       "\n",
       "                                             R2_ref  R2_start R2_MAPQ  \\\n",
       "E00218:136:HC35MCCXX:4:2117:4400:26062   MH533023.1        38      60   \n",
       "E00218:136:HC35MCCXX:6:1105:16083:63420  MH533023.1        86      40   \n",
       "E00218:136:HC35MCCXX:6:1116:6257:25025   MH533023.1       195      60   \n",
       "E00218:136:HC35MCCXX:2:1201:2299:47352   MH533023.1       438      60   \n",
       "E00218:136:HC35MCCXX:4:1213:2238:67780   MH533023.1       254      60   \n",
       "...                                             ...       ...     ...   \n",
       "E00218:136:HC35MCCXX:8:2224:32505:25939    unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:8:2224:32526:4596     unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:8:2224:32546:28787    unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:8:2224:32627:67006    unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:8:2224:32820:66883    unmapped  unmapped     N/A   \n",
       "\n",
       "                                        is_proper_pair  \n",
       "E00218:136:HC35MCCXX:4:2117:4400:26062            True  \n",
       "E00218:136:HC35MCCXX:6:1105:16083:63420           True  \n",
       "E00218:136:HC35MCCXX:6:1116:6257:25025            True  \n",
       "E00218:136:HC35MCCXX:2:1201:2299:47352            True  \n",
       "E00218:136:HC35MCCXX:4:1213:2238:67780            True  \n",
       "...                                                ...  \n",
       "E00218:136:HC35MCCXX:8:2224:32505:25939          False  \n",
       "E00218:136:HC35MCCXX:8:2224:32526:4596           False  \n",
       "E00218:136:HC35MCCXX:8:2224:32546:28787          False  \n",
       "E00218:136:HC35MCCXX:8:2224:32627:67006          False  \n",
       "E00218:136:HC35MCCXX:8:2224:32820:66883          False  \n",
       "\n",
       "[4025201 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paired_df = pd.DataFrame.from_dict(paired_alignment_dict, orient='index')\n",
    "paired_df.columns = ['R1_ref', 'R1_start', 'R1_MAPQ', 'R2_ref', 'R2_start', 'R2_MAPQ', 'is_proper_pair']\n",
    "paired_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_pairs_from_singleend_alignment(directory, refname=None):\n",
    "    \n",
    "   # check for alignment + original files and index files\n",
    "    try:\n",
    "        files = [glob.glob(os.path.join(directory, '*final.single.aln_all.bam'))[0],\n",
    "                 glob.glob(os.path.join(directory, '*final.improper_highq.bam'))[0],\n",
    "                 glob.glob(os.path.join(directory, '*final.map_unmap.bam'))[0]]\n",
    "\n",
    "    except:\n",
    "        print('One or more missing files!\\nExiting from get_pairs_from_singleend_alignment')\n",
    "        exit()\n",
    "    for file in files:\n",
    "        if not os.path.isfile(file+'.bai'):\n",
    "            pysam.samtools.index(file)\n",
    "        else:\n",
    "            if os.path.getctime(file+'.bai') < os.path.getctime(file):\n",
    "                pysam.samtools.index(file)\n",
    "    \n",
    "    output_dict = dict()\n",
    "    read_dict = defaultdict(lambda: [None, None])\n",
    "    \n",
    "    alignmentfile = pysam.AlignmentFile(files[0], 'rb')\n",
    "    duplicates = 0\n",
    "    for read in alignmentfile.fetch(refname, until_eof=True):\n",
    "        qname = read.query_name\n",
    "        # sanity check: no qname should have been in output_dict already\n",
    "        if qname in read_dict:\n",
    "            print('Duplicate read warning! {} is a duplicate'.format(qname))\n",
    "            duplicates += 1\n",
    "            continue\n",
    "        # single-end reads don't have read1/read2 tags\n",
    "        if read.is_unmapped:\n",
    "            read_dict[qname][0] = 'unmapped'\n",
    "        else:\n",
    "            read_dict[qname][0] = read\n",
    "    if duplicates > 0:\n",
    "        print('{} duplicate reads found in single-end alignment file. Skip and continue'.format(duplicates))\n",
    "\n",
    "    print('{} reads retrieved from single-end realignment'.format(len(read_dict)))\n",
    "    originalfiles = []\n",
    "    for file in files[1:]:\n",
    "        originalfile = pysam.AlignmentFile(file, 'rb')\n",
    "        originalfiles.append(originalfile)\n",
    "        for read in originalfile.fetch(refname, until_eof=True):\n",
    "            qname = read.query_name\n",
    "            if qname in read_dict:\n",
    "                if read.is_read1 and read_dict[qname][0] is not None:\n",
    "                    r2 = read_dict[qname][0]\n",
    "                    if read.is_unmapped:\n",
    "                        if r2 is 'unmapped':\n",
    "                            output_dict[qname] = ['unmapped', 'unmapped', 'N/A',\n",
    "                                                  r2, r2, 'N/A',\n",
    "                                                  'False']\n",
    "                        else:\n",
    "                            output_dict[qname] = ['unmapped', 'unmapped', 'N/A',\n",
    "                                                  alignmentfile.get_reference_name(r2.reference_id),\n",
    "                                                  r2.reference_start,\n",
    "                                                  r2.mapping_quality,\n",
    "                                                  'False']\n",
    "                    else:\n",
    "                        if r2 is 'unmapped':\n",
    "                            output_dict[qname] = [originalfile.get_reference_name(read.reference_id),\n",
    "                                                  read.reference_start,\n",
    "                                                  read.mapping_quality,\n",
    "                                                  r2, r2, 'N/A',\n",
    "                                                  'False']\n",
    "                        else:\n",
    "                            output_dict[qname] = [originalfile.get_reference_name(read.reference_id),\n",
    "                                                  read.reference_start,\n",
    "                                                  read.mapping_quality,\n",
    "                                                  alignmentfile.get_reference_name(r2.reference_id),\n",
    "                                                  r2.reference_start,\n",
    "                                                  r2.mapping_quality,\n",
    "                                                  read.is_proper_pair]\n",
    "                    # delete entry from dictionary as soon as its mate is found\n",
    "                    del read_dict[qname]\n",
    "                elif read.is_read2 and read_dict[qname][0] is not None:\n",
    "                    r1 = read_dict[qname][0]\n",
    "                    if read.is_unmapped:\n",
    "                        if r1 is 'unmapped':\n",
    "                            output_dict[qname] = [r1, r1, 'N/A',\n",
    "                                                  'unmapped', 'unmapped', 'N/A',\n",
    "                                                  'False']\n",
    "                        else:\n",
    "                            output_dict[qname] = [alignmentfile.get_reference_name(r1.reference_id), \n",
    "                                                  r1.reference_start,\n",
    "                                                  r1.mapping_quality,\n",
    "                                                  'unmapped', 'unmapped', 'N/A',\n",
    "                                                  'False']\n",
    "                    else:\n",
    "                        if r1 is 'low MAPQ' or r1 is 'unmapped':\n",
    "                            output_dict[qname] = [r1, r1, 'N/A',\n",
    "                                                  originalfile.get_reference_name(read.reference_id),\n",
    "                                                  read.reference_start,\n",
    "                                                  read.mapping_quality,\n",
    "                                                  'False']\n",
    "                        else:\n",
    "                            output_dict[qname] = [alignmentfile.get_reference_name(r1.reference_id), \n",
    "                                                  r1.reference_start,\n",
    "                                                  r1.mapping_quality,\n",
    "                                                  originalfile.get_reference_name(read.reference_id),\n",
    "                                                  read.reference_start,\n",
    "                                                  read.mapping_quality,\n",
    "                                                  r1.is_proper_pair]\n",
    "                    # delete entry from dictionary as soon as its mate is found\n",
    "                    del read_dict[qname]\n",
    "            # should only execute else when file == improper_highq.bam\n",
    "            else:\n",
    "                if 'map_unmap' in file:\n",
    "                    print('Warning: {} not found in realignment'.format(qname))\n",
    "                    \n",
    "                # if file == improper_highq.bam, also retrieve reads not found in read_dict\n",
    "                # as improperly paired high score reads\n",
    "                else:\n",
    "                    read_dict[qname][0] = read\n",
    "        originalfile.close()        \n",
    "\n",
    "    # sanity check: read_dict should be empty if all single-end reads found their mates\n",
    "    if bool(read_dict):\n",
    "        print('Pairing error! {} single-end reads do not have mates'.format(len(read_dict)))\n",
    "        for qname in read_dict:\n",
    "            print('Mate is not found for {}. Skip and continue'.format(qname))\n",
    "    \n",
    "    alignmentfile.close()\n",
    "    print('Done processing single-end alignment file!')\n",
    "    \n",
    "    return output_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4076492 single-end reads found\n",
      "Done processing single-end alignment file!\n"
     ]
    }
   ],
   "source": [
    "single_alignment_dict = get_pairs_from_singleend_alignment(os.path.join(directory, 'fam1/MH0143018'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>R1_ref</th>\n",
       "      <th>R1_start</th>\n",
       "      <th>R1_MAPQ</th>\n",
       "      <th>R2_ref</th>\n",
       "      <th>R2_start</th>\n",
       "      <th>R2_MAPQ</th>\n",
       "      <th>is_proper_pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:1:1217:17930:70980</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:4:1110:14296:56599</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:4:2112:26902:27802</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:5:2120:16580:38122</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>1</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:6:2106:13393:36786</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:1:1222:3710:30492</th>\n",
       "      <td>HLA-DQB1*03:03:02:02</td>\n",
       "      <td>644</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:3:2121:31308:37542</th>\n",
       "      <td>HLA-DQB1*03:05:01</td>\n",
       "      <td>5681</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:4:1202:30313:4948</th>\n",
       "      <td>HLA-DRB1*04:03:01</td>\n",
       "      <td>6848</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:7:2224:4481:6407</th>\n",
       "      <td>HLA-DRB1*08:03:02</td>\n",
       "      <td>5855</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:7:2123:26659:24954</th>\n",
       "      <td>HLA-DRB1*08:03:02</td>\n",
       "      <td>6525</td>\n",
       "      <td>36</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6125172 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       R1_ref R1_start  \\\n",
       "E00218:136:HC35MCCXX:1:1217:17930:70980                  chr1    10000   \n",
       "E00218:136:HC35MCCXX:4:1110:14296:56599                  chr1    10000   \n",
       "E00218:136:HC35MCCXX:4:2112:26902:27802                  chr1    10000   \n",
       "E00218:136:HC35MCCXX:5:2120:16580:38122                  chr1    10000   \n",
       "E00218:136:HC35MCCXX:6:2106:13393:36786                  chr1    10000   \n",
       "...                                                       ...      ...   \n",
       "E00218:136:HC35MCCXX:1:1222:3710:30492   HLA-DQB1*03:03:02:02      644   \n",
       "E00218:136:HC35MCCXX:3:2121:31308:37542     HLA-DQB1*03:05:01     5681   \n",
       "E00218:136:HC35MCCXX:4:1202:30313:4948      HLA-DRB1*04:03:01     6848   \n",
       "E00218:136:HC35MCCXX:7:2224:4481:6407       HLA-DRB1*08:03:02     5855   \n",
       "E00218:136:HC35MCCXX:7:2123:26659:24954     HLA-DRB1*08:03:02     6525   \n",
       "\n",
       "                                        R1_MAPQ    R2_ref  R2_start R2_MAPQ  \\\n",
       "E00218:136:HC35MCCXX:1:1217:17930:70980       0  unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:4:1110:14296:56599       0  unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:4:2112:26902:27802       0  unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:5:2120:16580:38122       1  unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:6:2106:13393:36786       0  unmapped  unmapped     N/A   \n",
       "...                                         ...       ...       ...     ...   \n",
       "E00218:136:HC35MCCXX:1:1222:3710:30492        0  unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:3:2121:31308:37542       0  unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:4:1202:30313:4948        0  unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:7:2224:4481:6407         0  unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:7:2123:26659:24954      36  unmapped  unmapped     N/A   \n",
       "\n",
       "                                        is_proper_pair  \n",
       "E00218:136:HC35MCCXX:1:1217:17930:70980          False  \n",
       "E00218:136:HC35MCCXX:4:1110:14296:56599          False  \n",
       "E00218:136:HC35MCCXX:4:2112:26902:27802          False  \n",
       "E00218:136:HC35MCCXX:5:2120:16580:38122          False  \n",
       "E00218:136:HC35MCCXX:6:2106:13393:36786          False  \n",
       "...                                                ...  \n",
       "E00218:136:HC35MCCXX:1:1222:3710:30492           False  \n",
       "E00218:136:HC35MCCXX:3:2121:31308:37542          False  \n",
       "E00218:136:HC35MCCXX:4:1202:30313:4948           False  \n",
       "E00218:136:HC35MCCXX:7:2224:4481:6407            False  \n",
       "E00218:136:HC35MCCXX:7:2123:26659:24954          False  \n",
       "\n",
       "[6125172 rows x 7 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "single_df = pd.DataFrame.from_dict(single_alignment_dict, orient='index')\n",
    "single_df.columns = ['R1_ref', 'R1_start', 'R1_MAPQ', 'R2_ref', 'R2_start', 'R2_MAPQ', 'is_proper_pair']\n",
    "single_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def generate_alignment_table(directory, mapq_threshold=10, refname=None):\n",
    "    \n",
    "    # create dictionary for pairs from realignment\n",
    "    alignment_dict = {**get_pairs_from_pairend_alignment(directory), **get_pairs_from_singleend_alignment(directory)}\n",
    "    print('Total: {} pairs of reads found from realignment'.format(alignment_dict))\n",
    "    \n",
    "    # check new alignment dict against original alignments and \n",
    "    # replace unmapped/low score alignments with original alignment\n",
    "    # check for alignment file and index file\n",
    "    try:\n",
    "        filename = glob.glob(os.path.join(directory, '*final.improper_lowq.bam'))[0]\n",
    "    except:\n",
    "        print('Alignment file not found!\\nExiting from generate_alignment_table')\n",
    "        exit()\n",
    "    if not os.path.isfile(filename+'.bai'):\n",
    "        pysam.samtools.index(filename)\n",
    "    else:\n",
    "        if os.path.getctime(filename+'.bai') < os.path.getctime(filename):\n",
    "            pysam.samtools.index(filename)\n",
    "    \n",
    "    alignmentfile = pysam.AlignmentFile(filename, 'rb')\n",
    "    count = 0\n",
    "    for read in alignmentfile.fetch(refname):\n",
    "        qname = read.query_name\n",
    "        if qname in alignment_dict:\n",
    "            if read.is_read1:\n",
    "                if alignment_dict[qname][0] == 'unmapped' or alignment_dict[qname][2] < mapq_threshold:\n",
    "                    alignment_dict[qname][:3] = [alignmentfile.get_reference_name(read.reference_id),\n",
    "                                                 read.reference_start,\n",
    "                                                 read.mapping_quality]\n",
    "            else:\n",
    "                if alignment_dict[qname][3] == 'unmapped' or alignment_dict[qname][5] < mapq_threshold:\n",
    "                    alignment_dict[qname][3:6] = [alignmentfile.get_reference_name(read.reference_id),\n",
    "                                                 read.reference_start,\n",
    "                                                 read.mapping_quality]\n",
    "        # sanity check: all reads should have been realigned and be found in alignment_dict\n",
    "        else:\n",
    "            count += 1\n",
    "            print('Warning: read {} not found in realignment. Skip and continue'.format(qname))\n",
    "        \n",
    "    if count > 0:\n",
    "        print('Warning: {} reads not found in realignment!')\n",
    "            \n",
    "    return alignment_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done processing paired-end alignment file!\n",
      "4076492 single-end reads found from single-end realignment\n",
      "Done processing single-end alignment file!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_data_rate_limit`.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4min 44s, sys: 1min 5s, total: 5min 49s\n",
      "Wall time: 7min 1s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "alignment_dict = generate_alignment_table(os.path.join(directory, 'fam1/MH0143018'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>R1_ref</th>\n",
       "      <th>R1_start</th>\n",
       "      <th>R1_MAPQ</th>\n",
       "      <th>R2_ref</th>\n",
       "      <th>R2_start</th>\n",
       "      <th>R2_MAPQ</th>\n",
       "      <th>is_proper_pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:4:2117:4400:26062</th>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>316</td>\n",
       "      <td>60</td>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>38</td>\n",
       "      <td>60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:6:1105:16083:63420</th>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>325</td>\n",
       "      <td>50</td>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>86</td>\n",
       "      <td>40</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:6:1116:6257:25025</th>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>432</td>\n",
       "      <td>60</td>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>195</td>\n",
       "      <td>60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:2:1201:2299:47352</th>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>43</td>\n",
       "      <td>60</td>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>438</td>\n",
       "      <td>60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:4:1213:2238:67780</th>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>454</td>\n",
       "      <td>60</td>\n",
       "      <td>MH533023.1</td>\n",
       "      <td>254</td>\n",
       "      <td>60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:1:1222:3710:30492</th>\n",
       "      <td>HLA-DQB1*03:03:02:02</td>\n",
       "      <td>644</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:3:2121:31308:37542</th>\n",
       "      <td>HLA-DQB1*03:05:01</td>\n",
       "      <td>5681</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:4:1202:30313:4948</th>\n",
       "      <td>HLA-DRB1*04:03:01</td>\n",
       "      <td>6848</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:7:2224:4481:6407</th>\n",
       "      <td>HLA-DRB1*08:03:02</td>\n",
       "      <td>5855</td>\n",
       "      <td>0</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E00218:136:HC35MCCXX:7:2123:26659:24954</th>\n",
       "      <td>HLA-DRB1*08:03:02</td>\n",
       "      <td>6525</td>\n",
       "      <td>36</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>unmapped</td>\n",
       "      <td>N/A</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10150373 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       R1_ref R1_start  \\\n",
       "E00218:136:HC35MCCXX:4:2117:4400:26062             MH533023.1      316   \n",
       "E00218:136:HC35MCCXX:6:1105:16083:63420            MH533023.1      325   \n",
       "E00218:136:HC35MCCXX:6:1116:6257:25025             MH533023.1      432   \n",
       "E00218:136:HC35MCCXX:2:1201:2299:47352             MH533023.1       43   \n",
       "E00218:136:HC35MCCXX:4:1213:2238:67780             MH533023.1      454   \n",
       "...                                                       ...      ...   \n",
       "E00218:136:HC35MCCXX:1:1222:3710:30492   HLA-DQB1*03:03:02:02      644   \n",
       "E00218:136:HC35MCCXX:3:2121:31308:37542     HLA-DQB1*03:05:01     5681   \n",
       "E00218:136:HC35MCCXX:4:1202:30313:4948      HLA-DRB1*04:03:01     6848   \n",
       "E00218:136:HC35MCCXX:7:2224:4481:6407       HLA-DRB1*08:03:02     5855   \n",
       "E00218:136:HC35MCCXX:7:2123:26659:24954     HLA-DRB1*08:03:02     6525   \n",
       "\n",
       "                                        R1_MAPQ      R2_ref  R2_start R2_MAPQ  \\\n",
       "E00218:136:HC35MCCXX:4:2117:4400:26062       60  MH533023.1        38      60   \n",
       "E00218:136:HC35MCCXX:6:1105:16083:63420      50  MH533023.1        86      40   \n",
       "E00218:136:HC35MCCXX:6:1116:6257:25025       60  MH533023.1       195      60   \n",
       "E00218:136:HC35MCCXX:2:1201:2299:47352       60  MH533023.1       438      60   \n",
       "E00218:136:HC35MCCXX:4:1213:2238:67780       60  MH533023.1       254      60   \n",
       "...                                         ...         ...       ...     ...   \n",
       "E00218:136:HC35MCCXX:1:1222:3710:30492        0    unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:3:2121:31308:37542       0    unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:4:1202:30313:4948        0    unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:7:2224:4481:6407         0    unmapped  unmapped     N/A   \n",
       "E00218:136:HC35MCCXX:7:2123:26659:24954      36    unmapped  unmapped     N/A   \n",
       "\n",
       "                                        is_proper_pair  \n",
       "E00218:136:HC35MCCXX:4:2117:4400:26062            True  \n",
       "E00218:136:HC35MCCXX:6:1105:16083:63420           True  \n",
       "E00218:136:HC35MCCXX:6:1116:6257:25025            True  \n",
       "E00218:136:HC35MCCXX:2:1201:2299:47352            True  \n",
       "E00218:136:HC35MCCXX:4:1213:2238:67780            True  \n",
       "...                                                ...  \n",
       "E00218:136:HC35MCCXX:1:1222:3710:30492           False  \n",
       "E00218:136:HC35MCCXX:3:2121:31308:37542          False  \n",
       "E00218:136:HC35MCCXX:4:1202:30313:4948           False  \n",
       "E00218:136:HC35MCCXX:7:2224:4481:6407            False  \n",
       "E00218:136:HC35MCCXX:7:2123:26659:24954          False  \n",
       "\n",
       "[10150373 rows x 7 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignment_df = pd.DataFrame.from_dict(alignment_dict, orient='index')\n",
    "alignment_df.columns = ['R1_ref', 'R1_start', 'R1_MAPQ', 'R2_ref', 'R2_start', 'R2_MAPQ', 'is_proper_pair']\n",
    "alignment_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "alignment_df.to_csv(os.path.join(directory, 'fam1/MH0143018/MH0143018.final_alignment_table.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fam 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done processing paired-end alignment file!\n"
     ]
    }
   ],
   "source": [
    "paired_alignment_dict = get_pairs_from_pairend_alignment(os.path.join(directory, 'fam3/03C16794'))\n",
    "paired_df = pd.DataFrame.from_dict(paired_alignment_dict, orient='index')\n",
    "paired_df.columns = ['R1_ref', 'R1_start', 'R1_MAPQ', 'R2_ref', 'R2_start', 'R2_MAPQ', 'is_proper_pair']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3071869 reads retrieved from single-end realignment\n",
      "Done processing single-end alignment file!\n"
     ]
    }
   ],
   "source": [
    "single_alignment_dict = get_pairs_from_singleend_alignment(os.path.join(directory, 'fam3/03C16794'))\n",
    "single_df = pd.DataFrame.from_dict(single_alignment_dict, orient='index')\n",
    "single_df.columns = ['R1_ref', 'R1_start', 'R1_MAPQ', 'R2_ref', 'R2_start', 'R2_MAPQ', 'is_proper_pair']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done processing paired-end alignment file!\n",
      "3071869 reads retrieved from single-end realignment\n",
      "Done processing single-end alignment file!\n",
      "Total: 7760642 pairs of reads found from realignment\n"
     ]
    }
   ],
   "source": [
    "alignment_dict = generate_alignment_dict(os.path.join(directory, 'fam3/03C16794'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleid = '03C16794'\n",
    "\n",
    "alignment_df = pd.DataFrame.from_dict(alignment_dict, orient='index')\n",
    "alignment_df.columns = ['R1_ref', 'R1_start', 'R1_MAPQ', 'R2_ref', 'R2_start', 'R2_MAPQ', 'is_proper_pair']\n",
    "\n",
    "alignment_df.to_csv(os.path.join(directory, 'fam3/03C16794', sampleid+'.final_alignment_table.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
